runtime/cudaq/qis/qubit_qis.h

```diff
@@ -18,661 +18,244 @@
-  };
-
-ConcreteQubitOp(h) ConcreteQubitOp(x) ConcreteQubitOp(y) ConcreteQubitOp(z)
-    ConcreteQubitOp(s) ConcreteQubitOp(t) ConcreteQubitOp(rx)
-        ConcreteQubitOp(ry) ConcreteQubitOp(rz) ConcreteQubitOp(r1)
-            ConcreteQubitOp(u3)
-
-} // namespace qubit_op
-
-// Convert a qubit to its unique id representation
-inline QuditInfo qubitToQuditInfo(qubit &q) { return {q.n_levels(), q.id()}; }
-inline bool qubitIsNegative(qubit &q) { return q.is_negative(); }
-
-#if CUDAQ_USE_STD20
-
-/// C++20 variant using templates.
-
-/// This function will apply the specified `QuantumOp`. It will check the
-/// modifier template type and if it is `base`, it will apply the operation to
-/// any qubits provided as input. If `ctrl`, it will take the first `N-1` qubits
-/// as the controls and the last qubit as the target.
-template <typename QuantumOp, typename mod = base, typename... QubitArgs>
-void oneQubitApply(QubitArgs &...args) {
-  // Get the name of this operation
-  auto gateName = QuantumOp::name();
-  static_assert(std::conjunction<std::is_same<qubit, QubitArgs>...>::value,
-                "Cannot operate on a qudit with Levels != 2");
-
-  // Get the number of input qubits
-  constexpr std::size_t nArgs = sizeof...(QubitArgs);
-
-  // Map the qubits to their unique ids and pack them.
-  std::vector<QuditInfo> quditInfos{qubitToQuditInfo(args)...};
-  std::vector<bool> qubitIsNegated{qubitIsNegative(args)...};
-
-  // If mod == base, then we just want to apply the gate to all qubits provided.
-  // This is a broadcast application.
-  if constexpr (std::is_same_v<mod, base>) {
-    for (auto &qubit : quditInfos)
-      getExecutionManager()->apply(gateName, {}, {}, {qubit});
-
-    // Nothing left to do, return
-    return;
-  }
+/// For C++17 we can't adhere to the language specification for
+/// the operation modifier type. For this case, we drop the modifier
+/// template parameter and users have access to a `cNAME` operation for
+/// single controlled operations.
+#ifdef CUDAQ_USE_STD20
+#define CUDAQ_MOD_TEMPLATE template <typename mod = base, typename... Args>
+#else
+#define CUDAQ_MOD_TEMPLATE template <typename... Args>
+#endif
 
-  // If we are here, then `mod` must be control or adjoint. Extract the controls
-  // and the target
-  std::vector<QuditInfo> controls(quditInfos.begin(),
-                                  quditInfos.begin() + nArgs - 1);
+namespace cudaq {
 
-  // If we have controls, check if any of them are negative controls, and if so
-  // apply an x.
-  if (!controls.empty())
-    for (std::size_t i = 0; i < controls.size(); i++)
-      if (qubitIsNegated[i])
-        getExecutionManager()->apply("x", {}, {}, {controls[i]});
+// --------------------------
+// Useful C++17 compliant concept checks (note we re-implement
+// std::remove_cvref since its a C++20 thing)
+template <typename T>
+using remove_cvref = std::remove_cv_t<std::remove_reference_t<T>>;
```

Comment:
```suggestion
#ifdef CUDAQ_USE_STD20
template <typename T>
using remove_cvref = std::remove_cvref<T>;
#else
template <typename T>
using remove_cvref = std::remove_cv_t<std::remove_reference_t<T>>;
#endif
```
makes it more explicit.